lib/linked_list.rb

```diff
@@ -18,128 +18,245 @@ def initialize
     def get_first
-      raise NotImplementedError
+      return nil if @head.nil?
+      return @head.data
     end
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(value)
-      raise NotImplementedError
+      @head.data = value if @head.nil?
+      current = @head
+      while current.next != nil
+        current = current.next
+      end
+      current.next = Node.new(value)
     end
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      return 0 if @head.nil?
+      count = 1
+      current = @head
+      while current.next != nil
+        current = current.next
+        count += 1
+      end
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      current_index = 0
+      current = @head
+      while current_index != index
+        return nil if current.nil? || current.next.nil?
+        current = current.next
+        current_index += 1
+      end
+      return current.data
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def visit
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      while current != nil || current.next != nil 
+        print current.data
+        current = current.next
+      end
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(value)
-      raise NotImplementedError
+      return nil if @head.nil?
+      current = @head
+      if @head.data == value
+        @head = @head.next
+      end
+
+      while current.next != nil
+        if current.next.data == value
+          current.next = current.next.next
+        end
+        current = current.next
+      end
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse
-      raise NotImplementedError
+      return nil if @head.nil?
+      return if @head.next.nil?
+
+      slow = @head
+      slow.next = nil
+      fast = @head.next
+      temp = fast.next

```

Comment:
```suggestion
      slow = @head
      fast = @head.next
      temp = fast.next
      slow.next = nil
```

Reply:
Does it make a difference? slow.next = nil does not change @head.next.